File: typescript/Services/SolutionRunner.ts
import type { ISolution } from "../Solutions/ISolution.ts";
import { SolutionOptions, SolutionType } from "../Models/SolutionModels.ts";

interface ISolutionRunner {
  run(): Promise<void>;
  runAll(): Promise<void>;
}

class SolutionRunner implements ISolutionRunner {
  public constructor(
    private readonly solutions: ISolution[],
    private readonly options: SolutionOptions,
  ) {}

  public async run(): Promise<void> {
    const solution = this.solutions.find((entry) => entry.day === this.options.day);

    if (!solution) {
      console.log(`Solution for day ${this.options.day} not found.`);
      return;
    }

    console.log(
      `Day ${this.options.day}, ${this.options.solutionType} part, ${this.options.runType} run`,
    );
    console.log(`https://adventofcode.com/2025/day/${this.options.day}`);
    console.log(`Solution: ${await solution.solve()}`);
  }

  public async runAll(): Promise<void> {
    const solutionsByDay = [...this.solutions].sort((a, b) => a.day - b.day);

    for (const solution of solutionsByDay) {
      this.options.solutionType = SolutionType.First;
      console.log(`Day: ${solution.day}, Solution 1: ${await solution.solve()}`);

      this.options.solutionType = SolutionType.Second;
      console.log(`Day: ${solution.day}, Solution 2: ${await solution.solve()}`);
    }
  }
}

export type { ISolutionRunner };
export { SolutionRunner };

File: dotnet/Services/SolutionRunner.cs
using AdventOfCode.Models;
using AdventOfCode.Solutions;
using Microsoft.Extensions.Options;

namespace AdventOfCode.Services;

public interface ISolutionRunner
{
    Task Run();
    Task RunAll();
}

public class SolutionRunner(IEnumerable<ISolution> solutions, IOptions<SolutionOptions> options) : ISolutionRunner
{
    public async Task Run()
    {
        var solution = solutions.FirstOrDefault(s => s.Day == options.Value.Day);
        if (solution == null)
        {
            Console.WriteLine($"Solution for day {options.Value.Day} not found.");
            return;
        }

        Console.WriteLine($"Day {options.Value.Day}, {options.Value.SolutionType} part, {options.Value.RunType} run");
        Console.WriteLine($"https://adventofcode.com/2025/day/{options.Value.Day}");
        Console.WriteLine($"Solution: {await solution.Solve()}");
    }

    public async Task RunAll()
    {
        var solutionsByDay = solutions.OrderBy(s => s.Day);

        foreach (var solution in solutionsByDay)
        {
            options.Value.SolutionType = SolutionType.First;
            Console.WriteLine($"Day: {solution.Day}, Solution 1: {await solution.Solve()}");

            options.Value.SolutionType = SolutionType.Second;
            Console.WriteLine($"Day: {solution.Day}, Solution 2: {await solution.Solve()}");
        }
    }
}